### PR TITLE
Add new "slugs.maxlength" option

### LINKED ISSUE #1821: Page title doesn't take OS' filepath length limit into account
You can create a page title with a very long name and this will result in an `mkdir(): No such file or directory` error. Obviously the maximum filepath length limit isn't taken into account while checking the title of a page.

Not sure how relevant this issue is - maybe it's relevant for API requests or virtual pages. I just spotted it by chance. Watch out for the length limits of Windows 10 and Linux, as they are different as you most likely know.

### PR DESCRIPTION
**Pull request by afbora**
## Describe the PR

Added new `slugs.maxlength` ~~content.slug.maxlength~~ option as @lukasbestle idea.
We can get your better ideas about option name and error message if you have.

## Related issues

- Fixes #1821 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Fixed code style issues with CS fixer and `composer fix`
- [x] Added in-code documentation (if needed)

### CODE REVIEW

**Comment by bastianallgeier** [comment]
This is great. I just wonder if we could somehow find a cleaner option name for it. We already have the slugs option. Ideally we would end up with something like: 

```php
return [
  'slugs' => [
    'language' => 'de',
    'maxlength' => 120
  ]
];
```

https://getkirby.com/docs/reference/system/options/slugs

What do you think?

**Comment by afbora** [comment]
Actually, the first thing that came to my mind was the `slugs` option, but I did not want to change the current process of `slugs` option. I like the `slugs.maxlength` option 👍

**Comment by bastianallgeier** [comment]
@afbora we would need to make sure though that `slugs => "en"` could still work.

**Comment by lukasbestle** [comment]
I agree with @bastianallgeier.

**Comment by afbora** [comment]
Great, done! 🎉 

I just wonder that following `slugs` option related with this? Please help me about that because i have no idea about system model:

https://github.com/getkirby/kirby/blob/master/config/api/models/System.php#L48-L50

**Comment by lukasbestle** [comment]
> I just wonder that following `slugs` option related with this? Please help me about that because i have no idea about system model:

No, that's just a value for the API, it's unrelated to the option.

File: src/Cms/PageRules.php [hunk 1]
```diff
@@ -293,6 +297,25 @@ public static function update(Page $page, array $content = []): bool
             ]);
         }
 
+        self::slugMaxlength($page);
+
         return true;
     }
+
+    protected static function slugMaxlength(Page $page, string $slug = null): void
```
Comment (lukasbestle) [inline src/Cms/PageRules.php]: ## 1. Naming

I think the name of the method should be `validateSlugLength` to make it clearer what it does.

## 2. Method signature

The second argument should be explicitly nullable:

```suggestion
    protected static function validateSlugLength(Page $page, ?string $slug = null): void
```

As the method doesn't always need the page (only if the slug is not given), the cleanest way would be to have only one argument, but type hinting both a page and a string is not possible until PHP 8.

Alternative: Only ever accept a string with the slug and make the other methods pass that explicitly:

```suggestion
    protected static function validateSlugLength(string $slug): void
```

I prefer the second alternative, but that's just personal preference.

## 3. Documentation

Could you also add a method docblock above it? All methods will have one at some point, so it's best to add them to new methods already while we still now what the method was supposed to do.

```php
/**
 * Ensures that the slug doesn't exceed the maximum length to make
 * sure that the directory name will be accepted by the filesystem
 *
 * @param string $slug New slug to check
 * @return void
 *
 * @throws \Kirby\Exception\InvalidArgumentException If the slug is too long 
 */
```

File: src/Cms/PageRules.php [hunk 2]
```diff
@@ -293,6 +297,25 @@ public static function update(Page $page, array $content = []): bool
             ]);
         }
 
+        self::slugMaxlength($page);
+
```
Comment (lukasbestle) [inline src/Cms/PageRules.php]: If I understand the logic correctly, updating the content won't change the slug, so this method wouldn't need to validate the slug length. Or have I missed something?

File: src/Cms/PageRules.php [hunk 3]
```diff
@@ -293,6 +297,25 @@ public static function update(Page $page, array $content = []): bool
...
         return true;
     }
+
+    protected static function slugMaxlength(Page $page, string $slug = null): void
+    {
+        if ($slugsMaxlength = option('slugs.maxlength')) {
```
Comment (lukasbestle) [inline src/Cms/PageRules.php]: I think there should be a default value, e.g. 150 characters. This default value should be used if the configured option is not set or if it's not already an integer.

So the `if` is not required, the check should always be made, either with the configured value or with the default.

File: tests/Cms/App/AppTest.php [hunk 4]
```diff
@@ -709,4 +709,45 @@ public function testVersionHash()
...
+
+        $this->assertSame('AE', Str::$language['Ä']);
+
+        // string option with dot notation
+        new App([
+            'roots' => [
+                'index' => '/dev/null'
+            ],
+            'options' => [
+                'slugs.language' => 'de'
```
Comment (lukasbestle) [inline tests/Cms/App/AppTest.php]: This second test should use a different language code to make sure that the first test isn't the reason why this passes.

File: src/Cms/PageRules.php (hunk 2 above)
Comment (afbora) [reply src/Cms/PageRules.php]: You right! I thought slug could be updated with the update() method, but this is not possible yes.

**Comment by afbora** [comment]
@lukasbestle Thank you for really helpful feedbacks. Implemented with default value is `255`

````php
$this->option('slugs', $this->option('slugs.language'))
````

I have to use above format to getting `slugs.language` option.
I know it's a bit complicated but you know that the user can make this definition in 3 different ways and we have a lack about dot notation options.

````php
return [
  'slugs' => 'de'
];

return [
  'slugs.language' => 'de'
];

return [
  'slugs' => [
    'language' => 'de'
  ]
];
````

**Comment by afbora** [comment]
@lukasbestle I used the `A::nest()` method in `App` class with 2 new places to access the different syntax settings defined at the init. Now what you think about that?

**Comment by lukasbestle** [comment]
I think we should wait for #2697. It’s likely that the way options are defined will be changed again.

I’m working on that issue and hope that I will have a solution this weekend.

**Comment by distantnative** [comment]
Could we add a few comments, especially to 
src/Cms/App.php 
src/Cms/AppTranslations.php 

Those parts never have been easy to read and understand. And these changes make it even more complex and difficult. Some comments would help a lot.
And also making some lines more explicit even if it makes the code overall a bit longer- but also a lot more easily to read (e.g. src/Cms/AppTranslations.php line 69 I think).

What do you think about this?

**Comment by lukasbestle** [comment]
@distantnative I have discussed with @afbora that we will add comments when reviewing this PR. I will take a look at the PR in detail in the next weeks.

**Comment by afbora** [comment]
@distantnative As you said, i added some annotations. Also phpDoc blocks have been improved in the classes i'm working on. @lukasbestle will go over 👍

**Comment by afbora** [comment]
Updated doc blocks according to @bastianallgeier says. Unit tests broken due to unneccasery space that related another PR and i've fixed on other PR.

**Review by bastianallgeier — APPROVED** [review]
LGTM!

**Pull request merged**